sklearn/utils/_readonly_array_wrapper.pyx

```diff
@@ -0,0 +1,64 @@
+# cython: language_level=3
+
+# -------------------------------------
+# Readonly array wrapper
+# -------------------------------------
+# TODO: Remove with Cython >= 3.0 which supports const memoryviews for fused types.
+#
+# This class supports the buffer protocol, thus can wrap arrays and memoryviews.
+# All it does is LIE about the readonly attribute: tell it's false!
+# This way, we can use it on arrays that we don't touch.
+# !!! USE CAREFULLY !!!
+
+
+from cpython cimport Py_buffer
+from cpython.buffer cimport PyObject_GetBuffer, PyBuffer_Release, PyBUF_WRITABLE
+
+import numpy as np
+cimport numpy as np
+
+
+np.import_array()
+
+
+ctypedef fused NUM_TYPES:
+    np.npy_float64
+    np.npy_float32
+    np.npy_int64
+    np.npy_int32
+
+
+cdef class ReadonlyWrapper:
```

Comment:
Should this be named more specifically? E.g. to match the naming of the file:

```suggestion
cdef class ReadonlyArrayWrapper:
```

Also, should we adapt the code-base to use it, e.g. [in KMeans](https://github.com/scikit-learn/scikit-learn/blob/8c4589b23c6481f978d4cfab511f25b77a805f13/sklearn/cluster/_k_means_lloyd.pyx#L5-L7)? This might make interfaces neater but at the expense of necessary boilerplate on arrays at downstream calls.

Reply:
Let's do KMeans (and maybe others) as follow-up.